core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoinExec.scala

```diff
@@ -183,17 +181,45 @@ case class HashJoinExec(leftKeys: Seq[Expression],
   // return empty here as code of required variables is explicitly instantiated
   override def usedInputs: AttributeSet = AttributeSet.empty
 
+  private def findShuffleDeps(buildRDDs: Seq[RDD[_]]) : Seq[Dependency[_]] = {
+    val buildShuffleDeps: Seq[Dependency[_]] = buildRDDs.flatMap(_.dependencies
+      .filter(_.isInstanceOf[ShuffleDependency[_, _, _]]))
+    val oneToOneRdds = buildRDDs.flatMap(_.dependencies).
+      filter(_.isInstanceOf[OneToOneDependency[_]]).map(_.rdd)
+    if (oneToOneRdds.isEmpty) {
+      buildShuffleDeps
+    } else {
+      buildShuffleDeps ++ findShuffleDeps(oneToOneRdds)
+    }
+  }
```

Comment:
As per our discussion on phone, point number 1 is required because the TakeOrderedAndProjectExec is introducing a ShuffledRowRDD over the replicated table. 

Refactored the code as per your comments for point 2.